src/supply_info.rs

```diff
@@ -106,80 +113,87 @@ mod tests {
         assert_eq!(supply_info.assets.len(), 0);
 
         // Add supply1
-        assert!(supply_info.add_supply(asset1, supply1).is_ok());
+        assert!(supply_info.add_supply(asset1, supply1.clone()).is_ok());
```

Comment:
Let's add `copy` for `AssetSupply` and remove `.clone()` everywhere.

Reply:
`AssetSupply` contains `HashSet` inside, which is not copyable, so `AssetSupply` can't be copyable either, but we can avoid usages of `.clone()` by pre-computing expected sum amounts at the beginning of the test and then use them in `asserts`, like this:

`rust
        let supply1 = AssetSupply::new(ValueSum::from_raw(20), false);
        let supply2 = AssetSupply::new(ValueSum::from_raw(30), true);
        let supply3 = AssetSupply::new(ValueSum::from_raw(10), false);
        let supply4 = AssetSupply::new(ValueSum::from_raw(10), true);
        let supply5 = AssetSupply::new(ValueSum::from_raw(50), false);

        let supply1_amount = sum([&supply1]).unwrap();
        let supply12_amount = sum([&supply1, &supply2]).unwrap();
        let supply123_amount = sum([&supply1, &supply2, &supply3]).unwrap();
        let supply1234_amount = sum([&supply1, &supply2, &supply3, &supply4]).unwrap();
        let supply5_amount = sum([&supply5]).unwrap();

        ...

        assert_eq!(
            supply_info.assets.get(&asset1),
            Some(&AssetSupply::new(supply1234_amount, true))
        );
`

In this case we don't need to clone `supply...` variables. By the way, if so, we can even remove that `sum` function and pre-compute sum amounts like this:

`
        let supply1 = AssetSupply::new(ValueSum::from_raw(20), false);
        let supply2 = AssetSupply::new(ValueSum::from_raw(30), true);
        let supply3 = AssetSupply::new(ValueSum::from_raw(10), false);
        let supply4 = AssetSupply::new(ValueSum::from_raw(10), true);
        let supply5 = AssetSupply::new(ValueSum::from_raw(50), false);

        let supply1_amount = supply1.amount;
        let supply12_amount = (supply1_amount + supply2.amount).unwrap();
        let supply123_amount = (supply12_amount + supply3.amount).unwrap();
        let supply1234_amount = (supply123_amount + supply4.amount).unwrap();
        let supply5_amount = supply5.amount;

        ...

        assert_eq!(
            supply_info.assets.get(&asset1),
            Some(&AssetSupply::new(supply1234_amount, true))
        );
`

What do you think?

Reply:
Get it now - I mixed it up with `SupplyInfo`. So, done: I added a `Copy` and removed `clone` calls.